scripts/helper_scripts/unipept-database-rs/src/bin/xml-parser.rs

```diff
@@ -0,0 +1,183 @@
+    Ok(())
+}
+
+type SmartStr = SmartString<LazyCompact>;
+
+#[derive(clap::ValueEnum, Clone, Debug)]
+enum UniprotType {
+    Swissprot,
+    Trembl,
+}
+
+// Parse a Uniprot XML file and convert it into a TSV-file
+#[derive(Parser, Debug)]
+struct Cli {
+    #[clap(value_enum, short = 't', long, default_value_t = UniprotType::Swissprot)]
+    uniprot_type: UniprotType,
+    #[clap(long, default_value_t = 0)]
+    threads: u32,
+    #[clap(short, long, default_value_t = false)]
+    verbose: bool,
+}
+
+fn write_header() {
+    let fields: [&str; 9] = [
+        "Entry",
+        "Sequence",
+        "Protein names",
+        "Version (entry)",
+        "EC number",
+        "Gene ontology IDs",
+        "Cross-reference (InterPro)",
+        "Status",
+        "Organism ID",
+    ];
+
+    let result_string = fields.join("\t");
+    println!("{}", result_string);
+}
+
+// Resolve the name of a single entry
+fn parse_name(entry: &uniprot::uniprot::Entry) -> SmartStr {
+    let mut submitted_name: SmartStr = SmartStr::new();
+
+    // Check the last "recommended" name from a protein's components,
+    // otherwise store the last "submitted" name of these components for later
+    for component in entry.protein.components.iter().rev() {
+        if let Some(n) = &component.recommended {
+            return n.full.clone();
+        }
+
+        if submitted_name.is_empty() {
+            if let Some(n) = component.submitted.last() {
+                submitted_name = n.full.clone();
+            }
+        }
+    }
+
+    // Do the same thing for the domains
+    for domain in entry.protein.domains.iter().rev() {
+        if let Some(n) = &domain.recommended {
+            return n.full.clone();
+        }
+
+        if submitted_name.is_empty() {
+            if let Some(n) = domain.submitted.last() {
+                submitted_name = n.full.clone();
+            }
+        }
+    }
+
+    // First check the protein's own recommended name,
+    // otherwise return the submitted name from above if there was one,
+    // otherwise the last submitted name from the protein itself
+    if let Some(n) = &entry.protein.name.recommended {
+        n.full.clone()
+    } else if !submitted_name.is_empty() {
+        submitted_name
+    } else if let Some(n) = entry.protein.name.submitted.last() {
+        n.full.clone()
+    } else {
+        eprintln!("Could not find a name for entry {}", entry.accessions[0]);
+        SmartStr::new()
+    }
+}
+
+// Write a single entry to stdout
+fn write_entry(entry: &uniprot::uniprot::Entry, verbose: bool) {
+    let accession_number: SmartStr = entry.accessions[0].clone();
+    let sequence: SmartStr = entry.sequence.value.clone();
+
+    let name: SmartStr = parse_name(entry);
+
+    let version: SmartStr = SmartStr::from(entry.version.to_string());
+
+    let mut ec_references: Vec<SmartStr> = Vec::new();
+    let mut go_references: Vec<SmartStr> = Vec::new();
+    let mut ip_references: Vec<SmartStr> = Vec::new();
```

Comment:
These can be `&str` rather than SmartStr, if you push references into them instead of clones on line 160.